tracer/test/Datadog.Trace.Security.IntegrationTests/AspNetBase.cs

```diff
@@ -117,19 +117,104 @@ public async Task TestBlockedRequestAsync(MockTracerAgent agent, string url, int
+            {
+                var start = DateTime.Now;
+                var nextBatch = start.AddSeconds(1);
+
+                await SendRequestsAsyncNoWaitForSpans(url, totalRequests);
+
+                var now = DateTime.Now;
+
+                if (now > nextBatch)
+                {
+                    // attempt to compensate for slow servers by increasing the error margin
+                    errorMargin *= 1.2;
+                    Console.WriteLine($"Failed to send all requests within a second now: {now:hh:mm:ss.fff}, nextBatch:{nextBatch:hh:mm:ss.fff}, error margin now: {errorMargin}");
+                }
+                else
+                {
+                    await Task.Delay(nextBatch - now);
+                }
+            }
+
+            var allSpansReceived = WaitForSpans(agent, iterations * totalRequests * spansPerRequest, "Overall wait", testStart);
+
+            var groupedSpans = allSpansReceived.GroupBy(s =>
+            {
+                var time = new DateTimeOffset((s.Start / TimeConstants.NanoSecondsPerTick) + TimeConstants.UnixEpochInTicks, TimeSpan.Zero);
+                return time.Second;
+            });
+
+            var spansWithUserKeep = allSpansReceived.Where(s =>
+            {
+                s.Tags.TryGetValue(Tags.AppSecEvent, out var appsecevent);
+                s.Metrics.TryGetValue("_sampling_priority_v1", out var samplingPriority);
+                return ((enableSecurity && appsecevent == "true") || !enableSecurity) && samplingPriority == 2.0;
+            });
+
+            var spansWithoutUserKeep = allSpansReceived.Where(s =>
+            {
+                s.Tags.TryGetValue(Tags.AppSecEvent, out var appsecevent);
+                return ((enableSecurity && appsecevent == "true") || !enableSecurity) && (!s.Metrics.ContainsKey("_sampling_priority_v1") || s.Metrics["_sampling_priority_v1"] != 2.0);
+            });
+            var itemsCount = allSpansReceived.Count();
+            var appsecItemsCount = allSpansReceived.Where(s =>
+            {
+                s.Tags.TryGetValue(Tags.AppSecEvent, out var appsecevent);
+                return appsecevent == "true";
+            }).Count();
+            if (enableSecurity)
+            {
+                var message = "approximate because of parallel requests";
+                var rateLimitOverPeriod = appsecTraceRateLimit * iterations;
+                if (appsecItemsCount >= rateLimitOverPeriod)
+                {
+                    var excess = appsecItemsCount - rateLimitOverPeriod;
+                    var spansWithUserKeepCount = spansWithUserKeep.Count();
+                    var spansWithoutUserKeepCount = spansWithoutUserKeep.Count();
+
+                    Console.WriteLine($"spansWithUserKeepCount: {rateLimitOverPeriod}, appsecTraceRateLimit: {rateLimitOverPeriod}");
+                    Console.WriteLine($"spansWithoutUserKeepCount: {spansWithoutUserKeepCount}, excess: {excess}");
```

Comment:
You can't use `Console.WriteLine` inside unit tests - you need to write them to `ITestOutputHelper`